lib/decoratex.ex

```diff
@@ -120,19 +120,42 @@ defmodule Decoratex do
       |> Post.decorate([:happy_comments_count, censured_comments: [pattern: list_of_words, replace: "*"]])
       ```
 
+  ## Reflection
+
+  Any decorated module will generate the `__decorate__` function that can be
+  used for runtime introspection of the decorations:
+
+  * `__decoration__(field)` - Returns the decortions of a field in format `%{type: type, function: function, options: options}`;
+
   """
 
-  defmacro __using__(_opts) do
+  @doc false
+  defmacro __using__(_) do
     quote do
-      import Decoratex
+      import Decoratex, only: [decorations: 1, add_decorations: 0]
     end
   end
 
+  @doc false
+  # credo:disable-for-next-line
+  # TODO: move decorate methods to another module and remove credo exception
+  # credo:disable-for-next-line
   defmacro decorations(do: block) do
     quote do
-      @decorations %{}
-      unquote(block)
-      def __decorations__, do: @decorations
+      Module.register_attribute(__MODULE__, :decorations, accumulate: true)
+
+      try do
+        import Decoratex
+        unquote(block)
+      after
+        :ok
+      end
+
+      decorations = @decorations |> Enum.reverse()
```

Comment:
I wouldn't use pipes if there is only one call

Reply:
Umm, i think that in some cases is more readable. Not this one but in the others (this line is from ecto 😅)